source/GameInterface/Services/Clans/Patches/ClanAddCompanionPatch.cs

```diff
@@ -0,0 +1,41 @@
+﻿using Common;
+using Common.Messaging;
+using Common.Util;
+using GameInterface.Services.Clans.Messages;
+using GameInterface.Services.GameDebug.Patches;
+using HarmonyLib;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.CampaignSystem.Actions;
+
+namespace GameInterface.Services.Clans.Patches
+{
+    [HarmonyPatch(typeof(AddCompanionAction), "ApplyInternal")]
+    public class ClanAddCompanionPatch
+    {
+        private static readonly AllowedInstance<Hero> AllowedInstance = new AllowedInstance<Hero>();
+
+        static bool Prefix(Clan clan, Hero companion)
+        {
+            CallStackValidator.Validate(companion, AllowedInstance);
+
+            if (AllowedInstance.IsAllowed(companion)) return true;
+
+            MessageBroker.Instance.Publish(clan, new AddCompanion(clan.StringId, companion.StringId));
```

Comment:
This should be CompanionAdded, this is an event, technically the state doesn't change or it is reacting to state that did change. This will be the same for all other events